tensorflow/core/kernels/data/cache_dataset_ops.cc

```diff
@@ -145,578 +275,641 @@ class CacheDatasetOp : public UnaryDatasetOpKernel {
-        // We intentionally use the same prefix for both `FileReaderIterator`
-        // and `FileWriterIterator`. Since at any time there will be at most
-        // one of them alive, there should be no conflicts. This allows both
-        // iterators to use a common key for `cur_index`. We leverage this
-        // in the corner case when this iterator is restored from an old
-        // checkpoint in `write` mode and the cache has been completely
-        // flushed to disk since then. In that case we simply build a
-        // `FileReaderIterator` and seek to the `cur_index`.
-        switch (mode_) {
-          case Mode::read:
-            iterator_ = absl::make_unique<FileReaderIterator>(
-                FileReaderIterator::Params{dataset(),
-                                           strings::StrCat(prefix(), "Impl")});
-            break;
-          case Mode::write:
-            iterator_ = absl::make_unique<FileWriterIterator>(
-                FileWriterIterator::Params{dataset(),
-                                           strings::StrCat(prefix(), "Impl")});
+        if (!reader_.Valid()) {
+          return errors::Internal("Error initializing BundleReader.");
         }
+        reader_.Seek(dataset()->FormatName(cur_index_, 0));
+        iterator_restored_ = true;
+        return Status::OK();
       }
 
+     private:
       mutex mu_;
-      enum Mode { read, write };
-      Mode mode_ GUARDED_BY(mu_);
-      std::unique_ptr<IteratorBase> iterator_ GUARDED_BY(mu_);
-    };  // FileIterator
-
-    const DatasetBase* const input_;
-    const string filename_;
-    Env* const env_;
-    const size_t num_tensors_;
-    const size_t tensor_index_padding_size_;
-    static const size_t kMaxItems = 10000000;  // 10 million
-    const size_t item_index_padding_size_;
-    const string tensor_format_string_;
-  };  // FileDataset
-
-  class MemoryDataset : public DatasetBase {
+      size_t cur_index_ GUARDED_BY(mu_);
+      BundleReader reader_ GUARDED_BY(mu_);
+      bool iterator_restored_ GUARDED_BY(mu_);
+    };  // FileReaderIterator
+
+    void InitializeIterator() EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+      // We intentionally use the same prefix for both `FileReaderIterator`
+      // and `FileWriterIterator`. Since at any time there will be at most
+      // one of them alive, there should be no conflicts. This allows both
+      // iterators to use a common key for `cur_index`. We leverage this
+      // in the corner case when this iterator is restored from an old
+      // checkpoint in `write` mode and the cache has been completely
+      // flushed to disk since then. In that case we simply build a
+      // `FileReaderIterator` and seek to the `cur_index`.
+      switch (mode_) {
+        case Mode::read:
+          iterator_ =
+              absl::make_unique<FileReaderIterator>(FileReaderIterator::Params{
+                  dataset(), strings::StrCat(prefix(), kImpl)});
```

Comment:
Note that this iterator prefix is not revised using `name_utils::IteratorPrefix()` as it is different from the prefix patterns we are using (there is no `::` and did not use dataset type as the prefix).